Implement Thread.findScopedValueBindings()

thallium opened the pull request:
Walk the stackframe to look for the runWith() method and get the first argument.
If the method is inlined, walk the oslots to find a ScopedValue$Snapshot instance.

thallium commented:
@fengxue-IS @babsingh @gacholio FYI

fengxue-IS commented on runtime/jcl/common/thread.cpp:
```diff
@@ -589,6 +589,44 @@ Java_java_lang_Thread_setExtentLocalCache(JNIEnv *env, jclass unusedClass, jobje
+				walkState->userData1 = (void *)arg;
+				rc = J9_STACKWALK_STOP_ITERATING;
+			} else {
+				walkState->flags |= J9_STACKWALK_ITERATE_O_SLOTS;
```
There should be a case to disable this flag up on seeing the next method or will this guarantee that result can be found in the oslot?

gacholio replied:
Ideally, once the appropriate frame/method is found, the walk should be stopped.

gacholio replied:
If you tag the "found" state in one of the walkState userData fields when the O-slot iterator detects the correct object, the next invocation of the frame iterator can detect this and abort the walk.

fengxue-IS commented on runtime/jcl/common/thread.cpp:
```diff
@@ -589,6 +589,44 @@ Java_java_lang_Thread_setExtentLocalCache(JNIEnv *env, jclass unusedClass, jobje
+		J9ROMMethod *romMethod = J9_ROM_METHOD_FROM_RAM_METHOD(method);
+		char *methodName = (char *)J9UTF8_DATA(J9ROMMETHOD_NAME(romMethod));
+
+		if (((J9VMJAVALANGTHREAD(vm) == methodClass)
+			|| (J9VMJAVALANGVIRTUALTHREAD(vm) == methodClass)
+			|| (J9VMJDKINCUBATORCONCURRENTSCOPEDVALUECARRIER(vm) == methodClass))
```
Does the vmconstantpool macro ensure `_OR_NULL` will print to the class when vm loaded it? I recall there was some issue with those limited to what is in the requiredClasses list

gacholio replied:
The `_OR_NULL` versions simply fetch the result from the constant pool with no resolution. They are mainly used for classes we know to have been pre-populated in the constant pool, i.e. any class in the constant pool that has fields or methods. In those cases, we don't check for the NULL result at all, assuming the slot is valid.

gacholio replied:
Fine for `Thread` and `VirtualThread` but not for `jdk/incubator/concurrent/ScopedValue$Carrier` since it declares no fields or method in the constant pool. Options are - declare a field, preload it manually during startup or handle the unresolved case here.

thallium replied:
I did encountered the case where I'm looking for an instance of `jdk/incubator/concurrent/ScopedValue$Carrier` but the `OR_NULL` version returns null.

gacholio commented on runtime/jcl/common/thread.cpp:
```diff
@@ -589,6 +589,44 @@ Java_java_lang_Thread_setExtentLocalCache(JNIEnv *env, jclass unusedClass, jobje
+		if (((J9VMJAVALANGTHREAD(vm) == methodClass)
+			|| (J9VMJAVALANGVIRTUALTHREAD(vm) == methodClass)
+			|| (J9VMJDKINCUBATORCONCURRENTSCOPEDVALUECARRIER(vm) == methodClass))
+			&& (0 == strncmp(methodName, "runWith", 7))
```
The macro also prevents overrunning short method names.

gacholio commented on runtime/jcl/common/thread.cpp:
```diff
@@ -602,8 +640,28 @@ Java_java_lang_Thread_setExtentLocalCache(JNIEnv *env, jclass unusedClass, jobje
+	walkState.walkThread = currentThread;
+	walkState.flags = J9_STACKWALK_ITERATE_FRAMES | J9_STACKWALK_MAINTAIN_REGISTER_MAP;
+	walkState.skipCount = 0;
+	walkState.userData1 = (void *)NULL;
```
Cast of NULL is never needed, in C or C++ afaik.

gacholio commented on runtime/jcl/common/thread.cpp:
```diff
@@ -589,6 +589,44 @@ Java_java_lang_Thread_setExtentLocalCache(JNIEnv *env, jclass unusedClass, jobje
+			|| (J9VMJDKINCUBATORCONCURRENTSCOPEDVALUECARRIER(vm) == methodClass))
+			&& (0 == strncmp(methodName, "runWith", 7))
+		) {
+			if (NULL != walkState->arg0EA) {
```
I believe that's true (excluding the end-of-stack marker frame, which you will not encounter with a visible frames walk).

gacholio replied:
This will not work for any JITted frame - without special handling, the contents of the arg0EA slot in compiled methods is unreliable. Perhaps this should operate under the same restrictions as `doPrivileged`? On the call we had discussed an alternate solution where the JCL code manually stacks these things to avoid the stack walk, but I'm not familiar with the precise APIs involved, so that might be impractical.

thallium commented:
Issues to be resolved:
- arg0EA is unreliable for JITted frames
- There could be multiple inlined `runWith()` and we can't tell which carrier in O slots is associated with the latest frame.

Since currently we don't have a test to verify the implementation, I'll just add an assertion (#17451) and we'll come back to this when the assertion is triggered.

babsingh commented on runtime/jcl/common/thread.cpp:
```diff
@@ -603,28 +603,94 @@ Java_java_lang_Thread_setExtentLocalCache(JNIEnv *env, jclass unusedClass, jobje
 /**
  * static native Object findScopedValueBindings();
  *
  * Find the most recent scoped value bindings in the stack.
```
```suggestion
 * Find the most recent scoped value binding in the stack.
```

babsingh reviewed (commented):
@thallium Also, post results of your personal build with the following test suites: sanity.functional, sanity.openjdk and extended.openjdk.

babsingh commented on runtime/jcl/common/thread.cpp:
```diff
@@ -603,28 +603,94 @@ Java_java_lang_Thread_setExtentLocalCache(JNIEnv *env, jclass unusedClass, jobje
 
 #if JAVA_SPEC_VERSION >= 20
+/**
+ * Stack walk function to find the most recent scoped value binding in the stack.
```
```suggestion
 * Stack frame iterator function to find the most recent scoped value binding in the stack.
```

babsingh commented on runtime/jcl/common/thread.cpp:
```diff
@@ -603,28 +603,94 @@ Java_java_lang_Thread_setExtentLocalCache(JNIEnv *env, jclass unusedClass, jobje
+			|| (J9VMJDKINCUBATORCONCURRENTSCOPEDVALUECARRIER_OR_NULL(vm) == methodClass))
+			&& J9UTF8_LITERAL_EQUALS(J9UTF8_DATA(methodName), J9UTF8_LENGTH(methodName), "runWith")
+		) {
+			walkState->flags |= J9_STACKWALK_ITERATE_O_SLOTS;
```
Add a comment here to highlight the issue and justify the approach taken. Something like ...

The goal is to find the first argument of the runWith method, which is an instance of jdk.incubator.concurrent.ScopedValue$Snapshot. But the first argument can't be reliably retrieved for any compiled method. The runWith method has two arguments, and there is only one argument which is an instance of jdk.incubator.concurrent.ScopedValue$Snapshot. So, all the arguments are looked at to find the Snapshot instance.

gacholio commented on runtime/jcl/common/thread.cpp:
```diff
@@ -603,28 +603,94 @@ Java_java_lang_Thread_setExtentLocalCache(JNIEnv *env, jclass unusedClass, jobje
+
+		if (((J9VMJAVALANGTHREAD_OR_NULL(vm) == methodClass)
+			|| (J9VMJAVALANGVIRTUALTHREAD_OR_NULL(vm) == methodClass)
+			|| (J9VMJDKINCUBATORCONCURRENTSCOPEDVALUECARRIER_OR_NULL(vm) == methodClass))
```
These are invalid - the OR_NULL macros can only be used for things that are guaranteed to be resolved (almost always at startup). Class refs are not resolve unless the constant pool also declared methods or fields related to the class.

gacholio commented:
I don't see anything here that addresses the JIT inlining issues we've previously discussed. Without that, this won't be reliable.

babsingh commented:
> I don't see anything here that addresses the JIT inlining issues we've previously discussed. Without that, this won't be reliable.

@0xdaryl had mentioned that there is special logic for `doPrivileged` to handle a similar scenario. But @tajila mentioned that the `runWith` methods will be reworked in the future. The easiest option will be to apply the `@DontInline` annotation to the `runWith` methods until they are reworked. Currently, the `runWith` methods have the `ForceInline` annotation, which will need to be removed. @thallium Can you confirm if `@DontInline` yields the correct behaviour?

gacholio commented:
To be completely correct, you also need to make sure that `runWith` also doesn't inline anything that would expose another instance of the bindings.

gacholio commented:
> @0xdaryl had mentioned that there is special logic for `doPrivileged` to handle a similar scenario.

Is there a reason not to extend this mechanism for `runWith` instead of the somewhat hacky solution here?

The optimization ensures that the relevant argument is backstored into the stack (and presumably kept alive in the stack map) so we can fetch it simply in the query.

0xdaryl commented:
> Is there a reason not to extend this mechanism for runWith instead of the somewhat hacky solution here?

I'll revisit that and get back to you here.

thallium commented:
The current assertion is triggered because the current scoped value binding of a newly-created thread is not correctly set ([detail](https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/649#issuecomment-1684098231)). A fix has been proposed and this PR can be closed for now.

babsingh commented:
@gacholio Can you please share the details of the workaround, which was mentioned in last Thursday's meeting?

gacholio commented:
A more sensible solution would be to modify the runWith methods to stack the bindings in a thread-local variable, but if we're not going to do that, then the following solution may work:

- Prevent JIT inlining of runWith
- Before walking the stack, set userData1 to NULL
- In the slot iterator, check each slot to see if it's an instance of the bindings. If it is, set userData1 to the object value (possibly asserting if the slot is already set).
- In the frame iterator, detect the runWith method. If found, check userData1. If it's filled in we're done (stop iterating), if it's not, that's probably another assert condition. If runWith is not found and inlineDepth in the walkState is 0, reset userData1 to NULL and continue iterating.

gacholio commented:
The inlining exception for runWith doesn't need to be absolute - we just have to ensure that we don't get more than one runWith in any single compilation.

@0xdaryl This seems like it should be relatively easy to do and would allow us proper inlining in the expected case (I'm not convinced the multiple runWith case can ever actually occur).

0xdaryl commented:
> This seems like it should be relatively easy to do and would allow us proper inlining in the expected case (I'm not convinced the multiple runWith case can ever actually occur).

Looking now...

nbhuiyan commented:
#18152 implements the JIT change that prevents inlining of `runWith`

gacholio commented:
> #18152 implements the JIT change that prevents inlining of `runWith`

This is probably good enough (considering this whole solution is a hack), but as dicussed the "correct" solution would be to ensure that only one instance of the bindings can appear in the stack frame of the compiled method.

babsingh commented:
The JIT change (https://github.com/eclipse-openj9/openj9/pull/18152) has been merged.

@thallium Can you please address the feedback from https://github.com/eclipse-openj9/openj9/pull/17402#issuecomment-1691889840 and any unaddressed feedback from the previous code review so that this PR can be merged?

thallium commented:
Applied the logic from https://github.com/eclipse-openj9/openj9/pull/17402#issuecomment-1691889840 and addressed all the feedbacks from @babsingh and @gacholio.

gacholio commented on runtime/jcl/common/thread.cpp:
```diff
@@ -603,28 +603,110 @@ Java_java_lang_Thread_setExtentLocalCache(JNIEnv *env, jclass unusedClass, jobje
+	jobject bindings = NULL;
+
+	walkState.walkThread = currentThread;
+	walkState.flags = J9_STACKWALK_ITERATE_FRAMES | J9_STACKWALK_MAINTAIN_REGISTER_MAP
```
Should add the VISIBLE_ONLY flag here which means the NULL check on the method in the iterator functions is unnecessary.

gacholio commented on runtime/jcl/common/thread.cpp:
```diff
@@ -603,28 +603,110 @@ Java_java_lang_Thread_setExtentLocalCache(JNIEnv *env, jclass unusedClass, jobje
+	J9JavaVM *vm = currentThread->javaVM;
+	j9object_t slotObject = *slot;
+	J9Class *clazz = J9OBJECT_CLAZZ(currentThread, slotObject);
+	J9Class *snapshotClazz = J9VMJAVALANGSCOPEDVALUESNAPSHOT_OR_NULL(vm);
```
This is invalid - the class may not be loaded, which will cause a crash.

thallium replied:
Should `J9VMJAVALANGSCOPEDVALUESNAPSHOT` be used instead?

gacholio replied:
I think that results in calling `findKnownClass` if the entry is not resolved. I think this is fine here, but it will need to be done while holding VM access (i.e. after entering the VM).

thallium replied:
Is this the same concern as mentioned in https://github.com/eclipse-openj9/openj9/pull/17402#discussion_r1294484802? I've added one field of `ScopedValue$Snapshot` to the constant pool.

gacholio commented on runtime/jcl/common/thread.cpp:
```diff
@@ -603,28 +603,110 @@ Java_java_lang_Thread_setExtentLocalCache(JNIEnv *env, jclass unusedClass, jobje
+		J9ROMMethod *romMethod = J9_ROM_METHOD_FROM_RAM_METHOD(method);
+		J9UTF8 *methodName = J9ROMMETHOD_NAME(romMethod);
+
+		if (((J9VMJAVALANGTHREAD_OR_NULL(vm) == methodClass)
```
Do all of the `runWith` have the same signature? If so, it might be better to do a name/sig text compare in order to avoid the OR_NULL issue of classes not being loaded.

gacholio replied:
It also catches any new additions without needing to change the code.

thallium replied:
> Do all of the `runWith` have the same signature?

No, there are two overloads in ScopedValue class: `runWith(Snapshot newSnapshot, Callable<R> op)` and `runWith(Snapshot newSnapshot, Runnable op)`.

gacholio replied:
Let me be more specific - do all the ones we're interested in have the same signature?

thallium replied:
Not quite sure what does "interested" mean here but I think those two overloads should be treated equally. Also, wouldn't the text comparison approach cause false positive (some other class have a runWith method with the same signature)?

gacholio replied:
Yes, I suppose that's technically possible.

gacholio commented on runtime/jcl/common/thread.cpp:
```diff
@@ -603,28 +603,110 @@ Java_java_lang_Thread_setExtentLocalCache(JNIEnv *env, jclass unusedClass, jobje
+			if (NULL != walkState->userData1) {
+				rc = J9_STACKWALK_STOP_ITERATING;
+			} else {
+				Assert_JCL_true(0);
```
This will assert any time we don't find the bindings. Probably best to hoist the userData1 check to the top of the function to avoid unncessary method comparisions when no binding have been found.

thallium replied:
Does that mean we will abandon this assertion? If we skip the method comparisons if we haven't find any bindings, we will miss the invalid case where we are in runWith method without finding a binding.

gacholio replied:
Yes, I think the assertion would be removed.

gacholio commented:
Once this works, we should be able to further optimize this by not setting the frame walk flag before the stack walk, and setting it in the slot iterator when bindings are found (and clear it again in the 0 inlineDepth case).

thallium commented:
@gacholio Hi, some of the feedbacks have been addressed. Can you take another look?

gacholio commented on runtime/jcl/common/thread.cpp:
```diff
@@ -603,28 +603,111 @@ Java_java_lang_Thread_setExtentLocalCache(JNIEnv *env, jclass unusedClass, jobje
+	 * We cannot load classes from inside the iterator callbacks, so we preload it
+	 * here and store them in the user data fields.
+	 */
+	walkState.userData2 = J9VMJAVALANGSCOPEDVALUESNAPSHOT_OR_NULL(vm);
```
These are still invalid (the OR_NULL will not load classes.

I'd prefer that all of the classes be loaded here and put into a struct that a single user data field points to (let's put all of the class loading assumptions in one place, rather than splitting the knowledge between the iterator and the caller). It's also somewhat self-documenting (there are no comments in the iterator around the use of the user data fields).

gacholio commented on runtime/jcl/common/thread.cpp:
```diff
@@ -603,28 +603,111 @@ Java_java_lang_Thread_setExtentLocalCache(JNIEnv *env, jclass unusedClass, jobje
+		if (NULL == walkState->userData1) {
+			walkState->userData1 = (void *)slotObject;
+		} else {
+			Assert_JCL_true(0);
```
This would read better with the assertion first, then the assignment to the slot (with some comments please).

gacholio commented on runtime/jcl/common/thread.cpp:
```diff
@@ -603,28 +603,111 @@ Java_java_lang_Thread_setExtentLocalCache(JNIEnv *env, jclass unusedClass, jobje
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	J9StackWalkState walkState = {0};
+	jobject bindings = NULL;
+
```
A description here of what's going on would be helpful (the fact that the slot iterator is called before the frame iterator).
